providers/implementations/kem/ml_kem-c.inc

```diff
@@ -0,0 +1,239 @@
+                             unsigned char *secret, size_t *secretlen)
+{
+    PROV_MLKEM_CTX *ctx = vctx;
+    MLKEM_KEY *key = ctx->key;
+    int ret;
+
+    if (key == NULL || key->keytype == 0) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return 0;
+    }
+
+    if (enc == NULL) {
+        if (enclen == NULL && secretlen == NULL)
+            return 0;
+        if (enclen != NULL)
+            *enclen = CIPHERTEXT_BYTES;
+        if (secretlen != NULL)
+            *secretlen = OSSL_MLKEM_SHARED_SECRET_BYTES;
+        return 1;
+    }
+
+    if (!ctx->have_entropy
+        && *secretlen < OSSL_MLKEM_SHARED_SECRET_BYTES) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*secretlen too small");
+        return 0;
+    }
+    if (*enclen < CIPHERTEXT_BYTES) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*enclen too small");
+        return 0;
+    }
+
+    if (!ctx->have_entropy && secret == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_SECRET);
+        return 0;
+    }
+
+    ret = (ctx->have_entropy) ?
+        encap_ext(enc, secret, pubptr(key), ctx->entropy,
+                  key->mlkem_ctx) :
+        encap_rng(enc, secret, pubptr(key), key->mlkem_ctx);
+
+    if (ctx->have_entropy) {
+        OPENSSL_cleanse(ctx->entropy, MLKEM_RANDOM_BYTES);
+        ctx->have_entropy = 0;
+    }
+    return ret;
+}
+
+static int mlkem_decapsulate(void *vctx,
+                             unsigned char *secret, size_t *secretlen,
+                             const unsigned char *enc, size_t enclen)
+{
+    PROV_MLKEM_CTX *ctx = vctx;
+    MLKEM_KEY *key = ctx->key;
+
+    if (key == NULL || key->keytype == 0) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return 0;
+    }
+
+    if (secret == NULL) {
+        *secretlen = OSSL_MLKEM_SHARED_SECRET_BYTES;
+        return 1;
+    }
+
+    if (*secretlen < OSSL_MLKEM_SHARED_SECRET_BYTES) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH, "*secretlen too small");
+        return 0;
+    }
+
+    /* MLKEM decap handles incorrect ciphertext lengths internally */
+    return decap((uint8_t *)secret, (uint8_t *)enc, enclen, prvptr(key),
+                 key->mlkem_ctx);
+}
+
+const OSSL_DISPATCH kemsym(ossl_mlkem,_asym_kem_functions)[] = {
+    { OSSL_FUNC_KEM_NEWCTX, (void (*)(void))mlkem_newctx },
+    { OSSL_FUNC_KEM_ENCAPSULATE_INIT,
+      (void (*)(void))mlkem_encapsulate_init },
+    { OSSL_FUNC_KEM_ENCAPSULATE, (void (*)(void))mlkem_encapsulate },
+    { OSSL_FUNC_KEM_DECAPSULATE_INIT,
+      (void (*)(void))mlkem_decapsulate_init },
+    { OSSL_FUNC_KEM_DECAPSULATE, (void (*)(void))mlkem_decapsulate },
+    { OSSL_FUNC_KEM_FREECTX, (void (*)(void))mlkem_freectx },
+    { OSSL_FUNC_KEM_SET_CTX_PARAMS,
+      (void (*)(void))mlkem_set_ctx_params },
+    { OSSL_FUNC_KEM_SETTABLE_CTX_PARAMS,
+      (void (*)(void))mlkem_settable_ctx_params },
+    OSSL_DISPATCH_END
+};
```

Comment:
Taking a closer look, but Is it worth the candle?
```
$ size providers/implementations/kem/lib*ml_kem*.o
   text    data     bss     dec     hex filename
   2577     240       0    2817     b01 providers/implementations/kem/libdefault-lib-ml_kem1024.o
   2577     240       0    2817     b01 providers/implementations/kem/libdefault-lib-ml_kem512.o
   2577     240       0    2817     b01 providers/implementations/kem/libdefault-lib-ml_kem768.o

$ size providers/implementations/keymgmt/libdefault-lib-*mlkem*.o
   text    data     bss     dec     hex filename
   6519     868       0    7387    1cdb providers/implementations/keymgmt/libdefault-lib-mlkem1024_kmgmt.o
   6519     868       0    7387    1cdb providers/implementations/keymgmt/libdefault-lib-mlkem512_kmgmt.o
   6519     868       0    7387    1cdb providers/implementations/keymgmt/libdefault-lib-mlkem768_kmgmt.o
```

Reply:
Refactored code, eliminating separate compilation for all but the thin top-level bindings of each variant. These are still needed because it is tricky To do the right thing portably in C90 while avoiding heap allocations inside the core algorithm (there are heap allocations for key storage. in the provider, and in various EVP functions, but inside the algorithm implementation all temporaries are on the stack).